daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/CommandPreprocessorSpec.scala

```diff
@@ -51,13 +71,120 @@ class CommandPreprocessorSpec extends AnyWordSpec with Matchers with TableDriven
   private[this] val compiledPackage = ConcurrentCompiledPackages()
   assert(compiledPackage.addPackage(defaultPackageId, pkg) == ResultDone.Unit)
 
+  val valueParties = ValueList(FrontStack(ValueParty("Alice")))
+
   "preprocessCommand" should {
 
+    val defaultPreprocesseor = new CommandPreprocessor(compiledPackage.interface, true, false)
+
+    "reject improperly typed commands" in {
+
+      val validCreate = CreateCommand(
+        "Mod:Record",
+        ValueRecord("", ImmArray("owners" -> valueParties, "data" -> ValueInt64(42))),
+      )
+      val validExe = ExerciseCommand(
+        "Mod:Record",
+        newCid,
+        "Transfer",
+        ValueRecord("", ImmArray("content" -> ValueList(FrontStack(ValueParty("Clara"))))),
+      )
+      val validExeByKey = ExerciseByKeyCommand(
+        "Mod:Record",
+        valueParties,
+        "Transfer",
+        ValueRecord("", ImmArray("content" -> ValueList(FrontStack(ValueParty("Clara"))))),
+      )
+      val validCreateAndExe = CreateAndExerciseCommand(
+        "Mod:Record",
+        ValueRecord("", ImmArray("owners" -> valueParties, "data" -> ValueInt64(42))),
+        "Transfer",
+        ValueRecord("", ImmArray("content" -> ValueList(FrontStack(ValueParty("Clara"))))),
+      )
+      val validFetch = FetchCommand(
+        "Mod:Record",
+        newCid,
+      )
+      val validFetchByKey = FetchByKeyCommand(
+        "Mod:Record",
+        valueParties,
+      )
+      val validLookup = LookupByKeyCommand(
+        "Mod:Record",
+        valueParties,
+      )
+      val negativeTestCases = Table[Command](
+        "command",
+        validCreate,
+        validExe,
+        validExeByKey,
+        validCreateAndExe,
+        validFetch,
+        validFetchByKey,
+        validLookup,
+      )
+
+      val positiveTestCases = Table[Command, ResultOfATypeInvocation[_]](
+        ("command", "error"),
+        validCreate.copy(templateId = "Mod:Undefined") ->
+          a[Error.Preprocessing.Lookup],
+        validCreate.copy(argument = ValueRecord("", ImmArray("content" -> ValueInt64(42)))) ->
+          a[Error.Preprocessing.TypeMismatch],
+        validExe.copy(templateId = "Mod:Undefined") ->
+          a[Error.Preprocessing.Lookup],
+        validExe.copy(choiceId = "Undefined") ->
+          a[Error.Preprocessing.Lookup],
+        validExe.copy(argument = ValueRecord("", ImmArray("content" -> ValueInt64(42)))) ->
+          a[Error.Preprocessing.TypeMismatch],
+        validExeByKey.copy(templateId = "Mod:Undefined") ->
+          a[Error.Preprocessing.Lookup],
+        validExeByKey.copy(contractKey = ValueList(FrontStack(ValueInt64(42)))) ->
+          a[Error.Preprocessing.TypeMismatch],
+        validExeByKey.copy(choiceId = "Undefined") ->
```

Comment:
```suggestion
        validExeByKey.copy(contractKey = ValueList(FrontStack(ValueInt64(42)))) ->
          a[Error.Preprocessing.TypeMismatch],
        validExeByKey.copy(choiceId = "Undefined") ->
```
nit: maybe consider swapping these two to be consistent with the other cases that first handle lookups and then type mismatches.

Reply:
ah makes sense, didn’t realize it was in field order :+1: